Add the Thrift tutorial (#4865)
line/armeria: site/src/pages/tutorials/thrift/blog/implement-delete.mdx

- `Main.java` for server
- `BlogServiceImpl.java` for server
- `BlogServiceExceptionHandler.java` for server
- `BlogClient.java` for client

## 1. Implement server-side

In the `BlogServiceImpl` class, implement the `deleteBlogPost()` method to delete a blog post.
Let's throw an exception in case there is no blog post for the given ID.

```java filename=BlogServiceImpl.java
@Override
public void deleteBlogPost(DeleteBlogPostRequest request, AsyncMethodCallback<Void> resultHandler)
        throws TException {
    final BlogPost removed = blogPosts.remove(request.getId());
    if (removed == null) {
        resultHandler.onError(
                new IllegalArgumentException("The blog post does not exist. ID: " + request.getId()));
    } else {
        resultHandler.onComplete(null);
    }
}
```

## 2. Implement client-side

In the `BlogClient` class, add the `deleteBlogPost()` method to send a request deleting a blog post.

```java filename=BlogClient.java
import example.armeria.blog.thrift.DeleteBlogPostRequest;

public class BlogClient {
    ...
    void deleteBlogPost(int id) throws Exception {
        final DeleteBlogPostRequest request = new DeleteBlogPostRequest().setId(id);

>         client.deleteBlogPost(request);

Reviewer: for consistency

```suggestion
        blogService.deleteBlogPost(request);
```

      throws TException {
  final BlogPost removed = blogPosts.remove(request.getId());
  if (removed == null) {
    resultHandler.onError(
            new IllegalArgumentException("The blog post does not exist. ID: " + request.getId()));
  } else {
    resultHandler.onComplete(null);
  }
}
```

## 2. Implement client-side

In the `BlogClient` class, add the `deleteBlogPost()` method to send a request deleting a blog post.

```java filename=BlogClient.java
import example.armeria.blog.thrift.DeleteBlogPostRequest;
...
void deleteBlogPost(int id) throws TException {
  final DeleteBlogPostRequest request = new DeleteBlogPostRequest().setId(id);
  blogService.deleteBlogPost(request);
}
```

## 3. Test deleting a blog post

Let's test deleting a blog post.
We'll delete the blog post with ID `1`, and try retrieving with the same ID to verify it is indeed deleted.
Add a test method like the following.

```java filename=BlogServiceTest.java
@Test
@Order(7)
void deleteBlogPost() throws TException {
  final BlogClient client = new BlogClient(server.httpUri(), "/thrift");
  client.deleteBlogPost(1);
  final Throwable exception = catchThrowable(() -> {
    client.getBlogPost(1);
  });

>   assertThat(exception).isInstanceOf(BlogNotFoundException.class)
>           .extracting("reason")
>           .asString()
>           .isEqualTo("The blog post does not exist. ID: 1");

Reviewer 2: What do you think about breaking line right before `.isInstanceOf()`? We could fix all similar sentences for aesthetics' sake:
```suggestion
  assertThat(exception)
    .isInstanceOf(BlogNotFoundException.class)
    .extracting("reason")
    .asString()
    .isEqualTo("The blog post does not exist. ID: 1");
```
Please note that I also reduced the indentation at minimum level so the code is readable even on mobile devices.

}
```

Run all the test cases on your IDE or using Gradle.
Check that you see the test is passed.

## 4. Test an error case

Let's test deleting a blog post that does not exist.
Add a test method like the following.

```java filename=BlogServiceTest.java
@Test
@Order(8)
void deleteInvalidBlogPost() {
  final BlogClient client = new BlogClient(server.httpUri(), "/thrift");
  final Throwable exception = catchThrowable(() -> {
    client.deleteBlogPost(100);
  });
  assertThat(exception).isInstanceOf(BlogNotFoundException.class)
                       .extracting("reason")
                       .asString()

>                        .isEqualTo("The blog post does not exist. ID: 100");

Reviewer 2: Ditto - please break more often and make them consistent.

             // See https://armeria.dev/docs/server-docservice for more information.
             .serviceUnder("/docs", docService)
             .build();
  }
  ```

3. (Optional) To access the Documentation service result easily, edit the log message in the `main()` method.
  ```java filename=Main.java highlight=2
  public static void main(String[] args) throws Exception {
    ...
    logger.info("Server has been started. Serving DocService at http://127.0.0.1:{}/docs",
                server.activeLocalPort());
  }
  ```
4. Now, [re-run the server](/tutorials/thrift/blog/run-server-and-client#2-run-the-server).

  The server and services are launched successfully if you see this message.
  ```bash
   Server has been started. Serving DocService at http://127.0.0.1:8080/docs
  ```

## 6. Check the DocService page

Let's test and call our service operations using Armeria's Documentation service.

1. Click the URL `http://127.0.0.1:8080/docs` from the log message or open up the URL on a web browser.

  If you see the Document service page, you've successfully launched the <type://DocService> and server.

  ![](../../../../images/tutorial_blogservice_thrift_docservice_start.png)

> 2. Click the **CreateBlogPost()** method link in the left panel. You can make calls to the method by clicking on the `Debug` button on the top right.

Reviewer: ```suggestion
2. Click the **createBlogPost()** method link in the left panel. You can make calls to the method by clicking on the `Debug` button on the top right.
```

> 2. Click the **CreateBlogPost()** method link in the left panel. You can make calls to the method from the page.

Reviewer: The most recent changes in `DocService` require that the `Debug` button be clicked to enter the debug menu.
<img width="1680" alt="Screen Shot 2023-05-16 at 6 22 40 PM" src="https://github.com/line/armeria/assets/8510579/b9b215fb-9418-48d6-9190-5cc6eb9c0d38">

I don't think we need a screenshot, but I think it would be nice if we can briefly mention it.

i.e.
```suggestion
2. Click the **CreateBlogPost()** method link in the left panel. You can make calls to the method by clicking on the `Debug` button on the top right.
```

    resultHandler.onError(
            new IllegalArgumentException("The blog post does not exist. ID: " + request.getId()));
  } else {
    resultHandler.onComplete(null);
  }
}
```

## 2. Implement client-side

In the `BlogClient` class, add the `deleteBlogPost()` method to send a request deleting a blog post.

```java filename=BlogClient.java
import example.armeria.blog.thrift.DeleteBlogPostRequest;
...
void deleteBlogPost(int id) throws TException {
  final DeleteBlogPostRequest request = new DeleteBlogPostRequest().setId(id);
  blogService.deleteBlogPost(request);
}
```

## 3. Test deleting a blog post

Let's test deleting a blog post.
We'll delete the blog post with ID `1`, and try retrieving with the same ID to verify it is indeed deleted.
Add a test method like the following.

```java filename=BlogServiceTest.java
@Test
@Order(7)
void deleteBlogPost() throws TException {
  final BlogClient client = new BlogClient(server.httpUri(), "/thrift");
  client.deleteBlogPost(1);
  final Throwable exception = catchThrowable(() -> {
    client.getBlogPost(1);
  });
  assertThat(exception)
    .isInstanceOf(BlogNotFoundException.class)
    .extracting("reason")

>     .asString()

Reviewer: ```suggestion
    .asString();
```

  ```java filename=Main.java highlight=5
  private static Server newServer(int port) throws Exception {
    ...
    return Server.builder()
             .http(port)
             .service("/thrift", tHttpService)
             // You can access the documentation service at http://127.0.0.1:8080/docs.
             // See https://armeria.dev/docs/server-docservice for more information.
             .serviceUnder("/docs", docService)
             .build();
  }
  ```

3. (Optional) To access the Documentation service result easily, edit the log message in the `main()` method.
  ```java filename=Main.java highlight=2
  public static void main(String[] args) throws Exception {
    ...
    logger.info("Server has been started. Serving DocService at http://127.0.0.1:{}/docs",
                server.activeLocalPort());
  }
  ```
4. Now, [re-run the server](/tutorials/thrift/blog/run-server#2-run-the-server).

  The server and services are launched successfully if you see this message.
  ```bash
   Server has been started. Serving DocService at http://127.0.0.1:8080/docs
  ```

## 6. Check the DocService page

Let's test and call our service operations using Armeria's Documentation service.

> 1. Click the URL `http://127.0.0.1:8080/docs` from the log message or open up the URL on a web browser.

Reviewer: So that users can just click on the hyperlink

```suggestion
1. Click the URL http://127.0.0.1:8080/docs from the log message or open up the URL on a web browser.
```
